### ISSUE TITLE
3D reconstruction from multiple images - scale mismatch in depth and pose

### ISSUE THREAD

**Issue by alexander-born** [TRI-ML/packnet-sfm#213]
Thanks for this great repository!

I tried to modify the camviz demo to not only create a pointcloud from a single image but from multiple images.

First I downloaded the pretrained KITTI checkpoints (PackNet01_HR_velsup_CStoK.ckpt).

I modified the [./scripts/infer.py](https://github.com/TRI-ML/packnet-sfm/blob/master/scripts/infer.py#L87) to also calculate the pose like this:
```py
# pose inference
pose = model_wrapper.pose(image, [image_tminus1, image_tplus1]) 
```
and saved these relative poses from (t -> t_minus_1) additionally to the depth in a npz file.

To create a [Pose](https://github.com/TRI-ML/camviz/blob/ba1d8b92dff2ae01616c999739eca3e507434f4f/camviz/objects/pose.py#L21) object from the pose net output I created a transformation matrix with:
```py
from scipy.spatial.transform import Rotation as R
def pose_vec2mat(posenet_output): #posenet_output is the pose from t -> t_minus_1 
    """Convert Euler parameters to transformation matrix."""
    trans, rot = posenet_output[:3], posenet_output[3:]
    rot_mat = R.from_euler("zyx", rot).as_matrix()
    mat = np.concatenate((rot_mat, trans[:, np.newaxis]), axis=1)  # [B,3,4]
    padding = np.array([0, 0, 0, 1])
    mat = np.concatenate((mat, padding[np.newaxis, :]), axis=0)
    return mat

```

I accumulated these poses with [`__matmul__`](https://github.com/TRI-ML/camviz/blob/master/camviz/objects/pose.py#L186) to get all the camera poses. Is the pose calculation correct (it looks good when visualized)?

Then I used [`Camera.i2w()`](https://github.com/TRI-ML/camviz/blob/master/camviz/objects/camera.py#L161)
 to project the point clouds of multiple images to world coordinates (Additionally I filtered the point cloud by a max. distance threshold).

It seems like there is a scale mismatch in the output of the depth network and the pose network. This can be seen in the screenshot below, where I am visualizing pointclouds from multiple images (KITTI Tiny). The coordinate systems in the screenshot are the camera poses. You can see the scale mismatch in the duplicated vehicle and that the camera poses moved way too much compared to the pointcloud. Shouldn't these pointclouds of multiple images fit together really good, when using the pose and depth net which are trained together?
![image](https://user-images.githubusercontent.com/61082865/153902935-8bd8caf8-b443-4830-b268-ae18e8ac1c27.png)

Only if I either scale depth or poses the resulting pointclouds are overlapping (not perfectly): (scaled pose with factor 0.14, no scaling factor in depth):
![image](https://user-images.githubusercontent.com/61082865/153918584-50016e16-e5b4-4700-a4a3-42e36f38af36.png)
Another kitti example (scaled pose with factor 0.14, no scaling factor in depth):
![image](https://user-images.githubusercontent.com/61082865/153917526-81423405-74e1-4d49-99e1-c19b4a6987ac.png)

**Comment by hhhharold** [comment]
Have you tried to 3D reconstruction from multi-view images? Just like the image shown in the README of the camviz

**Comment by alexander-born** [comment]
Yes this is a 3D reconstruction from multiple views in a global frame. I generated one joint 3d point cloud from multiple images and joined them together via the relative poses between the images.

In the camviz readme gifs something else is done. There only the point clouds from single image are shown on after the other in the camera frame and not in a global frame. 

The problem I am facing is that the scale of the pose net output does not match the scale of the depth net.

**Comment by VitorGuizilini-TRI** [comment]
Can you try using ground-truth information, just to check whether the transformations are correct?

**Comment by hhhharold** [comment]
> Can you try using ground-truth information, just to check whether the transformations are correct?

Yes, there was something mistakes in my code and the extrinsics is correct. By the way, how to set the parameter of pose in draw.add3Dworld function to get a better original visual angle in world coordinates. The default setting in demo is draw.add3Dworld('wld', luwh=(0.33, 0.00, 1.00, 1.00), pose=(7.25323, -3.80291, -5.89996, 0.98435, 0.07935, 0.15674, 0.01431)).

**Comment by sctrueew** [comment]
Hi everyone, I have a video like this, and I've extracted depth and converted each frame to a points cloud. I would like to display it like this (SLAM). Could you please guide me?

![image](https://github.com/user-attachments/assets/e3e60d61-450b-4d38-9960-316d3a13027f)
